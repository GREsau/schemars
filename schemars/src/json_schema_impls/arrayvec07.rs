use crate::gen::SchemaGenerator;
use crate::schema::*;
use crate::JsonSchema;
use arrayvec07::{ArrayString, ArrayVec};
use std::convert::TryInto;

// Do not set maxLength on the schema as that describes length in characters, but we only
// know max length in bytes.
forward_impl!((<const CAP: usize> JsonSchema for ArrayString<CAP>) => String);

impl<T, const CAP: usize> JsonSchema for ArrayVec<T, CAP>
where
    T: JsonSchema,
{
    no_ref_schema!();

    fn schema_name() -> String {
        format!("Array_up_to_size_{}_of_{}", CAP, T::schema_name())
    }

    fn json_schema(gen: &mut SchemaGenerator) -> Schema {
        SchemaObject {
            instance_type: Some(InstanceType::Array.into()),
            array: Some(Box::new(ArrayValidation {
                items: Some(gen.subschema_for::<T>().into()),
                max_items: CAP.try_into().ok(),
                ..Default::default()
            })),
            ..Default::default()
        }
        .into()
    }
}
