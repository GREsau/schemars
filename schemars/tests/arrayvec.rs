mod util;
use util::*;

#[test]
fn arrayvec07() -> TestResult {
    test_default_generated_schema::<arrayvec07::ArrayVec<i32, 16>>("arrayvec")
}

#[test]
fn arrayvec07_string() -> TestResult {
    test_default_generated_schema::<arrayvec07::ArrayString<16>>("arrayvec_string")
}
