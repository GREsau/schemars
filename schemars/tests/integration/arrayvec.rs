use crate::prelude::*;
use arrayvec07::{ArrayString, ArrayVec};

#[test]
fn arrayvec07() {
    test!(ArrayVec<i32, 8>)
        .assert_snapshot()
        .assert_allows_ser_roundtrip([
            ArrayVec::from_iter([]),
            ArrayVec::from_iter([1, 2, 3, 4, 5, 6, 7, 8]),
        ])
        .assert_matches_de_roundtrip(
            (0..16).map(|len| Value::Array((0..len).map(Value::from).collect())),
        )
        .assert_matches_de_roundtrip(arbitrary_values_except(
            is_array_of_u64,
            "FIXME schema allows out-of-range positive integers",
        ));
}

#[test]
fn arrayvec07_arraystring() {
    test!(ArrayString<8>)
        .assert_identical::<String>()
        .assert_allows_ser_roundtrip(["".try_into().unwrap(), "12345678".try_into().unwrap()])
        .assert_matches_de_roundtrip(arbitrary_values_except(
            Value::is_string,
            "There's not a good way to express UTF-8 byte length in JSON schema, so schema ignores the ArrayString's capacity.",
        ));
}

fn is_array_of_u64(value: &Value) -> bool {
    value
        .as_array()
        .is_some_and(|a| a.iter().all(Value::is_u64))
}
