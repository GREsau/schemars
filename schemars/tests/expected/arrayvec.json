{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "title": "Array_up_to_size_16_of_int32",
  "type": "array",
  "items": {
    "type": "integer",
    "format": "int32"
  },
  "maxItems": 16
}